### ISSUE TITLE
stdlib.com an easy way to deploy to AWS Lambda + API Gateway?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-aws-lambda#93]
Keith Horwood of http://stdlib.com/ reached out for feedback.
![image](https://cloud.githubusercontent.com/assets/194400/22394818/aa3d6ee6-e522-11e6-8727-c989b312473a.png)
Since he asked nicely (_and I believe in giving `before` you get..._)
I'm giving him ***One Pomodoro*** of my time. ⌛️ 💸 😉 

First impression is: ***Really promising***!
![image](https://cloud.githubusercontent.com/assets/194400/22394839/6c6ce924-e523-11e6-89e6-7be15ed4102f.png)

Automatic Scaling Without Configuration
![image](https://cloud.githubusercontent.com/assets/194400/22394855/d0c2d104-e523-11e6-9624-24c53e5a3eeb.png)
> ***How***? _given that Lambda has an initial **limit of 100 _concurrent_ invocations** ... 
which you then have to "**contact support**" to get extended according to your needs..._
see: http://stackoverflow.com/questions/36826352/aws-lambda-toomanyrequestsexception-rate-exceeded (_I **know**, because I've **felt** the **pain** of hitting the "**Lambda-Limit**"..._) 😉 
Is the `stdlib.com` service running my lambdas on _their_ AWS account in which case they are handling the scaling for me? so I _don't_ have to _manually_ contact AWS support...?

Ease of deployment looks great!
![image](https://cloud.githubusercontent.com/assets/194400/22394861/e3ddeff8-e523-11e6-9d6b-a45cba6c7373.png)

A Complete End-to-End "Server-less" Solution
![image](https://cloud.githubusercontent.com/assets/194400/22394868/f4ce2e7c-e523-11e6-927c-a81952d627f3.png)

GitHub Repo: clearly got some good traction. ⭐️ 
![image](https://cloud.githubusercontent.com/assets/194400/22394824/fc8b00aa-e522-11e6-99fe-ebd400d53a83.png)

Docs look great! http://docs.stdlib.com/main/#/introduction ✅ 
![image](https://cloud.githubusercontent.com/assets/194400/22394829/278adf0a-e523-11e6-9807-ecfd80945da9.png)

Signed up (_though not a fan of a **modal** for signup, I'm sure they have A/B tested it for conversion..._)
![image](https://cloud.githubusercontent.com/assets/194400/22394900/6bf47460-e525-11e6-8c0c-97bc2469f0d7.png)

Next action after signup is _unclear_... what do I do now? 
![image](https://cloud.githubusercontent.com/assets/194400/22394901/7ebf47be-e525-11e6-8938-dd34169b82f1.png)
Will reading the docs get me **up and running in the next _minute_** `before` I **get _distracted_** by a notification?

Don't see the point of a profile page that I have to _manually_ update...
![image](https://cloud.githubusercontent.com/assets/194400/22394917/ba022030-e525-11e6-95bd-f68cf5f73e2b.png)
I would have used GitHub auth here... get people's details and profile pic from GitHub... 👤 
or Twitter. most Devs I know have twitter... 🐦 

Scrolled down and see that I have **No Services Available**...
![image](https://cloud.githubusercontent.com/assets/194400/22394935/44cf0548-e526-11e6-9b9c-a6c65c5efd29.png)

I would have a link to help the person create their "***Hello World***" first service.
e.g: http://docs.stdlib.com/main/#/quickstart
![image](https://cloud.githubusercontent.com/assets/194400/22394954/baa75784-e526-11e6-9790-e0f65b4a831f.png)

The next _logical_ thing for me after I'm "_sold_" on the ***features*** is to ask "***How Much***?!"
Pricing: http://stdlib.com/pricing is where the site lets me down. still _unclear_... 🤔 
![image](https://cloud.githubusercontent.com/assets/194400/22394826/14fb933e-e523-11e6-9984-b5063a9bfb31.png)
This is the only thing "_holding me back_" from building something on it ***today***.
I like to know ***exactly*** how something can be "***free***". 
Does Keith have a mountain of money he's burning to get his service to critical mass and then sell it to ***Big Company XYZ*** (_e.g: MSFT who aren't competing particularly well in against AWS Lambda..._)

## Feedback to Keith

+ [x] ***Website*** (Landing Page) is **superb**. 
  + [x] **clear offering** and call-to-action to try the product!
+ [x] Docs are great and able to get up-and-running quickly!

### `TODO`

In order of importance to _me_ (_a potential user_):

+ [ ] Pricing holding me back from adopting it... ***how*** is it "***free***"? (_VC Funded?_)
will it have the same _fate_ as [Nodejitsu](https://news.ycombinator.com/item?id=9033175) ?
I don't _want_ to invest my time in learning/using a (_admittedly great looking_) platform for the "_rug_" to be pulled in a year's time and be _forced_ to "_migrate_" all my microservices at short-notice... clarify this set up a basic PayPal form to accept $10 from me an I will try it. Being able to ***pay*** to use a service actually gives it _credibility_!! 👍 
+ [ ] "***Try Online***" deploy the hello world example (_without any "global" installation_)
+ [ ] How do I ***run locally*** and what is the "**_Continuous Integration_ Story**"...?
+ [ ] Do I still have to pay AWS or does `stdlib` handle the payment to AWS (_and then is there a markup for the Compute/Gateway resources or same cost_?)
+ [ ] Give me a way to make a "***Buildpack***" https://devcenter.heroku.com/articles/buildpacks which I can ***single-click deploy*** a microservice to AWS and I'm _SOLD_! 

# Conclusion

Sadly my ***Pomodoro*** is has run out! ⌛️ 
but I would _invest_ more time in this if these "Todo" items were addressed by Keith and his team! 
I _cannot_ recommend something I haven't _tried_. (_and nobody `else` should!!_)
Looks really promising.
@keithwhor is a _credible_ developer with _great_ project: https://github.com/keithwhor/nodal under his belt so I ***believe*** this has ***real potential***! `#WatchThisSpace`!

<br /> <br /> <br /> <br />

# tl;dr

## There Are _Many_ Serverless "Solutions" ... what is the stdlib `USP`?

***Why*** would I use this _instead_ of 
+ AWS Lambda (_the Gorilla_!)
+ https://serverless.com (_the "works everywhere" market leading framework_...)
+ https://github.com/claudiajs/claudia deployment tool to Lambda + Gateway
+ https://github.com/Miserlou/Zappa (_credible Python-based system with more traction than stdlib_)

> see discussion: https://news.ycombinator.com/item?id=13089630

## Security Concerns 

> _**Note**: this is a _personal_ "**bugbear**" of mine and not a direct "**criticism**" of anyone_ @stdlib.com

![why-install-globally](https://cloud.githubusercontent.com/assets/194400/22394995/d2eec6f0-e527-11e6-84d5-eb54c5d83e11.png)

***Why*** do I _need_ to install the CLI globally? 
***why*** should I ***trust*** a random package from NPM with ***global*** access to my system? Can't I `try` this _without_ making a ***commitment*** of installing a CLI **globally**?

> _to be clear I - and most people - tend to give people the benefit of the doubt when it comes to the security of their code, but `should` I...?!?_
`this` _**could** be the **ultimate engineering attack!!**..._ @evilpacket ***right***...? 💭 
how **trivial** would it be to setup a _nice-looking website_ buy some adwords for "Serverless" or "mine" GitHub for email addresses and get devs to install a "toolkit" that is actually a [**rootkit**](https://en.wikipedia.org/wiki/Rootkit) ...?

# What are _your_ impressions?

**Comment by keithwhor** [comment]
Hey Nelson --- thanks for making this thread for feedback.

I actually want to be clear, we run our own gateway and we don't consider "deploying to Lambda" a core competency. It's something we do, but we think about deploying to Lambda like changing your own brakes --- if it's really "zero-ops", why do I still feel like I need trustworthy mechanics? (serverless, apex, claudia, etc.) We, in a sense, take "serverless" for granted - it's clearly the future of compute, so what's next? That's what we're focused on - the next abstractions. Deployment is trivial, we can all write `.zip` tools and call an API. (Though we're still happy about our workflow.)

It's best if you think of what we're doing with StdLib almost like GitHub for service deployment. We'll be publishing more information in the coming weeks about pricing (we've hammered most of it down). I don't want to use this as an excuse, because execution and delivery are ultimately all that matters, but if you want to know why we're going to stick around --- the majority of the workflows you see were created by one developer - that's me - with the web UIs generated by a very talented Google engineer who joined our team only a couple of months ago. There's still a lot to build, and though we are venture-backed, we're also extremely capital efficient (we're pretty proud of that --- having your own compute platform that  you can dogfood endlessly helps in a bit of a virtuous cycle). The team we've brought together is extremely competent, and we're lucky to have them.

Thanks again for this --- the criticism is extremely helpful and we'll be executing on it as fast as we can. We're far from perfect, and we work within tight constraints to deliver as much value as possible. We'd love it if you and others let us know what you want to see next --- because we're building it. :)

Thanks so much,
- Keith

**Comment by keithwhor** [comment]
Also, if you wanna check out a recent article we wrote about massively parallel DNA sequence alignment on StdLib, here it is :) (I dropped out of an MS in Biochemistry to run my first startup, hence the little bit of specialization in the space):

https://hackernoon.com/using-server-less-architecture-to-massively-parallelize-dna-sequence-alignment-via-stdlib-and-343dd2d5aebf#.p0my37b7i

**Comment by nelsonic** [comment]
@keithwhor I'm _really_ excited for what you & your team are building! I can see from the commit history that you have made a _monumental_ effort over the last few months and if you continue your momentum it will be an _amazing_ product! I will fire up a VM later and install your CLI _globally_ and give it a shot. The fact that you replied on GitHub gives me a _lot_ of _faith_ that you aren't some "customer service" rep who "values my call", but rather are a passionate and dedicated CEO who is _executing_ on a vision for what "serverless" _can_ be!
I will gladly do a _full_ write-up "post" on `stdlib` when pricing is available. 
But _meanwhile_, if you can assure me that it will be in a _similar_ ballpark to Lambda+Gateway, I'm happy to write a service and deploy it to your system `this` weekend...
Thanks again for reaching out and for sharing the link(s)!
From everyone @dwyl, all the very best for 2017! ❤️  ✅  🚀

**Comment by keithwhor** [comment]
I promise that it will be the easiest way you've deployed a function as a web API. 🙂 

Expect pricing to be out and solidified soon. I will definitely post an update here when we're ready! Thanks so much for the feedback, we live and die by our developers, so I want to make sure everybody's voice is heard.

**Comment by nelsonic** [comment]
Closing as this is no longer relevant. 📕 
The domain https://stdlib.com/ is ***parked***. 🅿️ 

<img width="894" height="615" alt="Image" src="https://github.com/user-attachments/assets/7c64fc0d-8719-4b7d-8fde-17e396c16d28" />

Not redirected to a rebranded project page ... mothballed. 🫠
